conn.go

```diff
@@ -0,0 +1,460 @@
+func (c *Conn) Close() error {
+	return c.close(nil)
+}
+
+// Call initiates a JSON-RPC call using the specified method and
+// params, and waits for the response. If the response is successful,
+// its result is stored in result (a pointer to a value that can be
+// JSON-unmarshaled into); otherwise, a non-nil error is returned.
+func (c *Conn) Call(ctx context.Context, method string, params, result interface{}, opts ...CallOption) error {
+	call, err := c.DispatchCall(ctx, method, params, opts...)
+	if err != nil {
+		return err
+	}
+	return call.Wait(ctx, result)
+}
+
+// DisconnectNotify returns a channel that is closed when the
+// underlying connection is disconnected.
+func (c *Conn) DisconnectNotify() <-chan struct{} {
+	return c.disconnect
+}
+
+// DispatchCall dispatches a JSON-RPC call using the specified method
+// and params, and returns a call proxy or an error. Call Wait()
+// on the returned proxy to receive the response. Only use this
+// function if you need to do work after dispatching the request,
+// otherwise use Call.
+func (c *Conn) DispatchCall(ctx context.Context, method string, params interface{}, opts ...CallOption) (Waiter, error) {
+	req := &Request{Method: method}
+	for _, opt := range opts {
+		if opt == nil {
+			continue
+		}
+		if err := opt.apply(req); err != nil {
+			return Waiter{}, err
+		}
+	}
+	if err := req.SetParams(params); err != nil {
+		return Waiter{}, err
+	}
+	call, err := c.send(ctx, &anyMessage{request: req}, true)
+	if err != nil {
+		return Waiter{}, err
+	}
+	return Waiter{call: call}, nil
+}
+
+// Notify is like Call, but it returns when the notification request
+// is sent (without waiting for a response, because JSON-RPC
+// notifications do not have responses).
+func (c *Conn) Notify(ctx context.Context, method string, params interface{}, opts ...CallOption) error {
+	req := &Request{Method: method, Notif: true}
+	for _, opt := range opts {
+		if opt == nil {
+			continue
+		}
+		if err := opt.apply(req); err != nil {
+			return err
+		}
+	}
+	if err := req.SetParams(params); err != nil {
+		return err
+	}
+	_, err := c.send(ctx, &anyMessage{request: req}, false)
+	return err
+}
+
+// Reply sends a successful response with a result.
+func (c *Conn) Reply(ctx context.Context, id ID, result interface{}) error {
+	resp := &Response{ID: id}
+	if err := resp.SetResult(result); err != nil {
+		return err
+	}
+	_, err := c.send(ctx, &anyMessage{response: resp}, false)
+	return err
+}
+
+// ReplyWithError sends a response with an error.
+func (c *Conn) ReplyWithError(ctx context.Context, id ID, respErr *Error) error {
+	_, err := c.send(ctx, &anyMessage{response: &Response{ID: id, Error: respErr}}, false)
+	return err
+}
+
+// SendResponse sends resp to the peer. It is lower level than (*Conn).Reply.
+func (c *Conn) SendResponse(ctx context.Context, resp *Response) error {
+	_, err := c.send(ctx, &anyMessage{response: resp}, false)
+	return err
+}
+
+func (c *Conn) close(cause error) error {
```

Comment:
I changed the order of the methods slightly by making sure that public methods come before private methods.